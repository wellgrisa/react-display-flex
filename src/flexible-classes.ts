export const flexibleBoxClasses: { [key in FlexibleBoxClasses]: string } = {
  alignContentBaseline: 'align-content-baseline',
  alignContentCenter: 'align-content-center',
  alignContentEnd: 'align-content-end',
  alignContentSpaceAround: 'align-content-space-around',
  alignContentSpaceBetween: 'align-content-space-between',
  alignContentSpaceEvenly: 'align-content-space-evenly',
  alignContentStart: 'align-content-start',
  alignContentStretch: 'align-content-stretch',
  alignItemsBaseline: 'align-items-baseline',
  alignItemsCenter: 'align-items-center',
  alignItemsEnd: 'align-items-end',
  alignItemsStart: 'align-items-start',
  alignItemsStretch: 'align-items-stretch',
  alignSelfAuto: 'align-self-auto',
  alignSelfBaseline: 'align-self-baseline',
  alignSelfCenter: 'align-self-center',
  alignSelfEnd: 'align-self-end',
  alignSelfStart: 'align-self-start',
  alignSelfStretch: 'align-self-stretch',
  column: 'react-display-flex-column',
  columnReverse: 'react-display-flex-column-reverse',
  flex: 'react-display-flex-flex',
  flexGrow: 'react-display-flex-grow',
  inlineFlex: 'react-display-flex-inline-flex',
  justifyContentCenter: 'justify-content-center',
  justifyContentEnd: 'justify-content-end',
  justifyContentSpaceAround: 'justify-content-space-around',
  justifyContentSpaceBetween: 'justify-content-space-between',
  justifyContentSpaceEvenly: 'justify-content-space-evenly',
  justifyContentStart: 'justify-content-start',
  noWrap: 'react-display-flex-no-wrap',
  row: 'react-display-flex-row',
  rowReverse: 'react-display-flex-row-reverse',
  wrap: 'react-display-flex-wrap',
  wrapReverse: 'react-display-flex-wrap-reverse',
}

export type FlexibleBoxClasses =
  | 'alignItemsStart'
  | 'alignItemsEnd'
  | 'alignItemsCenter'
  | 'alignItemsBaseline'
  | 'alignItemsStretch'
  | 'justifyContentStart'
  | 'justifyContentEnd'
  | 'justifyContentCenter'
  | 'justifyContentSpaceAround'
  | 'justifyContentSpaceEvenly'
  | 'justifyContentSpaceBetween'
  | 'alignContentStart'
  | 'alignContentEnd'
  | 'alignContentCenter'
  | 'alignContentSpaceAround'
  | 'alignContentSpaceEvenly'
  | 'alignContentSpaceBetween'
  | 'alignContentBaseline'
  | 'alignContentStretch'
  | 'alignSelfStart'
  | 'alignSelfEnd'
  | 'alignSelfCenter'
  | 'alignSelfBaseline'
  | 'alignSelfStretch'
  | 'alignSelfAuto'
  | 'column'
  | 'row'
  | 'columnReverse'
  | 'rowReverse'
  | 'wrap'
  | 'noWrap'
  | 'wrapReverse'
  | 'flex'
  | 'flexGrow'
  | 'inlineFlex'
