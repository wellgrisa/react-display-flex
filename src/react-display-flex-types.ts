import { PropsWithChildren } from 'react'

type AlignItemsType = {
  alignItemsStart?: boolean
  alignItemsEnd?: boolean
  alignItemsCenter?: boolean
  alignItemsBaseline?: boolean
  alignItemsStretch?: boolean
}

type JustifyContentType = {
  justifyContentStart?: boolean
  justifyContentEnd?: boolean
  justifyContentCenter?: boolean
  justifyContentSpaceAround?: boolean
  justifyContentSpaceEvenly?: boolean
  justifyContentSpaceBetween?: boolean
}

type AlignContentType = {
  alignContentStart?: boolean
  alignContentEnd?: boolean
  alignContentCenter?: boolean
  alignContentSpaceAround?: boolean
  alignContentSpaceEvenly?: boolean
  alignContentSpaceBetween?: boolean
  alignContentBaseline?: boolean
  alignContentStretch?: boolean
}
type AlignSelfPropType = {
  alignSelfStart?: boolean
  alignSelfEnd?: boolean
  alignSelfCenter?: boolean
  alignSelfBaseline?: boolean
  alignSelfStretch?: boolean
  alignSelfAuto?: boolean
}

type FlexDirectionType = {
  column?: boolean
  row?: boolean
  columnReverse?: boolean
  rowReverse?: boolean
}

type FlexGrowType = {
  flexGrow?: boolean
}

type FlexibleHTMLElements = 'article' | 'aside' | 'div' | 'figure' | 'footer' | 'header' | 'main' | 'nav' | 'section'

type ElementType = {
  element?: FlexibleHTMLElements
}

type DisplayType = {
  flex?: boolean
  inlineFlex?: boolean
}

type WrapType = {
  wrap?: boolean
  noWrap?: boolean
  wrapReverse?: boolean
}

type HTMLAttributesType = React.HTMLAttributes<HTMLElement>

export type ReactDisplayFlexPropTypes = AlignItemsType &
  JustifyContentType &
  AlignContentType &
  AlignSelfPropType &
  FlexDirectionType &
  FlexGrowType &
  ElementType &
  DisplayType &
  WrapType

export type ReactDisplayFlexPropTypesWithHtmlAttributes = PropsWithChildren<
  ReactDisplayFlexPropTypes & HTMLAttributesType
>
