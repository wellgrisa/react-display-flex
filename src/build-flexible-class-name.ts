import { flexibleBoxClasses } from './flexible-classes'
import { FlexibleBoxPropTypes } from './flexible-types'

export const buildFlexibleClassName = ({ display, className, ...props }: FlexibleBoxPropTypes) =>
  Object.keys(props)
    .filter((key) => key in flexibleBoxClasses)
    .reduce(
      (classes, property) => `${classes} ${flexibleBoxClasses[property] ? flexibleBoxClasses[property] : ''}`,
      `${className ? `${className} ` : ''}flexible-box${
        flexibleBoxClasses[display] ? ` ${flexibleBoxClasses[display]}` : ''
      }`,
    )
