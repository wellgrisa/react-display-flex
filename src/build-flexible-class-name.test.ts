import { buildFlexibleClassName } from './build-flexible-class-name'
import { flexibleBoxClasses } from './flexible-classes'
import { ReactDisplayFlexPropTypes, ReactDisplayFlexPropTypesWithHtmlAttributes } from './react-display-flex-types'

describe('Building the flexible class name', () => {
  it('returns the flexible class name', () => {
    const singleFlexibleClassProperty: ReactDisplayFlexPropTypes = {}

    const className = buildFlexibleClassName(singleFlexibleClassProperty)

    expect(className).toBe('react-display-flex')
  })

  it('returns the given className given', () => {
    const className = buildFlexibleClassName({
      className: 'given-class-name',
    })

    expect(className).toBe('given-class-name react-display-flex')
  })

  it('returns the flexible class name with a single flex class property', () => {
    const singleFlexibleClassProperty: ReactDisplayFlexPropTypes = {
      alignContentStart: true,
    }

    const className = buildFlexibleClassName(singleFlexibleClassProperty)

    expect(className).toBe('react-display-flex align-content-start')
  })

  it('returns the flexible class name ignoring html attributes', () => {
    const singleFlexibleClassProperty: ReactDisplayFlexPropTypesWithHtmlAttributes = {
      alignContentStart: true,
      role: 'dialog',
    }

    const className = buildFlexibleClassName(singleFlexibleClassProperty)

    expect(className).toBe('react-display-flex align-content-start')
  })

  it('returns the flexible class name with every property contained in the flexible classes object', () => {
    const multipleFlexProperties = Object.keys(flexibleBoxClasses).reduce((flexibleProperties, flexibleProperty) => {
      flexibleProperties[flexibleProperty] = true

      return flexibleProperties
    }, {}) as ReactDisplayFlexPropTypes

    const className = buildFlexibleClassName(multipleFlexProperties)

    expect(className).toMatchSnapshot()
  })
})
