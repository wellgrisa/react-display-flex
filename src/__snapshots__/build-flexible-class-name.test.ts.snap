// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Building the flexible class name returns the flexible class name with every property contained in the flexible classes object 1`] = `"react-display-flex align-content-baseline align-content-center align-content-end align-content-space-around align-content-space-between align-content-space-evenly align-content-start align-content-stretch align-items-baseline align-items-center align-items-end align-items-start align-items-stretch align-self-auto align-self-baseline align-self-center align-self-end align-self-start align-self-stretch react-display-flex-column react-display-flex-column-reverse react-display-flex-flex react-display-flex-grow react-display-flex-inline-flex justify-content-center justify-content-end justify-content-space-around justify-content-space-between justify-content-space-evenly justify-content-start react-display-flex-no-wrap react-display-flex-row react-display-flex-row-reverse react-display-flex-wrap react-display-flex-wrap-reverse"`;
